### PR TITLE
INTEL MKL: Enhance MkL BatchNorm ops with primitive reuse

### PR DESCRIPTION
**Pull request by gzmkl**
Enable MKL BatchNorm ops with primitive reuse, to improve
  (1) model training and
  (2) inference of small batch size
by minimizing primitive creation time.

************ Notes *******************
Please review and merge this PR first
 https://github.com/tensorflow/tensorflow/pull/19399

### CODE REVIEW

**Comment by gzmkl** [comment]
Close temporarily - pending conv_bwd PR

**Comment by gzmkl** [comment]
mkl_conv_ops.cc has been reverted to avoid any review confusion. 
The new version of mkl_conv_ops.cc is contained in PR #19399. 

Thanks

**Comment by gzmkl** [comment]
Pending on https://github.com/tensorflow/tensorflow/pull/19754

**Comment by gzmkl** [comment]
Reopen since PR #19399 has been merged

**Comment by yiqianglee** [comment]
Yes, i have already singed CLA, and should be ok to contribute.

**Comment by gzmkl** [comment]
Hi, 
I am the PR submitter and confirm that I am OK with all the changes made by yiqianglee.
Thanks,
GZ

**Comment by rmlarsen** [comment]
@gzmkl I think many of my comments for PR 19403 would apply to this as well, please modify accordingly.

**Comment by gzmkl** [comment]
Plan to apply  PR 19403 comments on this PR too.

**Comment by gzmkl** [comment]
Latest code change based on PR 19403 code review suggestions.
Summary:
   (1) Change signature of Execute (...) so that input parameters are declared as "const" while output parameter as "non-const". 
   (2) In many places, remove unnecessary static_cast and const_cast to simplify the code.
   (3) Back-out code in SetOp() method (mkl_util.h) - using "emplace" causes unit test failures of ops/nn_grad_test.py. 
   (4) Minor changes in some comment statements to make descriptions more accurate. 
BTW, related code review suggestions from PR 19399 & 19400 have also been reflected in this PR.

**Comment by rmlarsen** [comment]
@gzmkl Thanks for the update. Let's try to get PR 19403 merged first before we push the remaining PRs. Running tests for it again now.

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc [hunk 1]
```diff
@@ -1330,8 +1706,10 @@ class MklFusedBatchNormGradOp : public OpKernel {
     dnn_shape_diff_src.SetMklTensor(false);
     AllocateOutputSetMklShape(context, kDiffSrcIndex, diff_src_tensor,
                               tf_shape_src, dnn_shape_diff_src);
-    for (size_t i = 0; i < (*diff_src_tensor)->shape().num_elements(); i++)
-      (*diff_src_tensor)->flat<T>().data()[i] = 0;
+    int num_elements = (*diff_src_tensor)->shape().num_elements();
+    auto diff_src_data = (*diff_src_tensor)->flat<T>().data();
+    for (size_t i = 0; i < num_elements; i++)
```
Comment (rmlarsen) [inline tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: I would use std::fill instead of a loop. Same everywhere.

**Review by rmlarsen — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc (hunk 1 above)
Comment (gzmkl) [reply tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Sure, I will do. 
There are places in files which are not changed in this PR. Should I do the same change?

Comment (rmlarsen) [reply tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Please do. :-)

**Comment by rmlarsen** [comment]
@gzmkl resolved conflict

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc (hunk 1 above)
Comment (gzmkl) [reply tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: I have done code refactoring for multiple places in mkl_fused_batch_norm_op.cc. 

To avoid potential merge conflicts, I did not apply this code change recommendation to
source files not related to this PR. 

We will include this suggestion, along with others (such as changing signature of Execute()
with proper const or non-const argument declarations), to create a separate "code clean up"
PR. 

Thanks!

**Review by rmlarsen — APPROVED** [review]

**Comment by gzmkl** [comment]
Hi Rasmus,

Please choose "master" version to address the following conflict in mkl_util.h

  <<<<<<< primreuse_batch_norm
#include <vector>
=======
>>>>>>> master

Line of "#include <vector>" seems to be duplicated.

**Comment by gzmkl** [comment]
Please take the branch code with the following conflict
<<<<<<< primreuse_batch_norm
#include "tensorflow/core/platform/cpu_info.h"                          // Keep this line
=======
>>>>>>> master

Thanks!

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc [hunk 2]
```diff
@@ -262,6 +262,7 @@ class MklFusedBatchNormOp : public OpKernel {
     }
 
     void MklCreateInputLayout(OpKernelContext* context) {
+      const Tensor& input = MklGetInput(context, 0);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Just curious, why do we have this line? This local variable isn't used anywhere in the function.

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc [hunk 3]
```diff
@@ -544,6 +545,7 @@ class MklFusedBatchNormGradOp : public OpKernel {
...
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: This local variable also isn't used anywhere in this function.

Comment (gzmkl) [reply tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Good catch!  

Yes, this line of code should be there. And it was in a function with MKL-ML integration which will be removed in the long run since we have MKL-DNN integration.

I will clean up this code, after all primitive reuse PRs have been done (only Relu one remains). I have a TODO list based on Rasmus's suggestions, which were applied only to individual PRs  (thus
only on related changed files), and will have a "clean-up" PR. 

Thanks!

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc (hunk 2 above)
Comment (gzmkl) [reply tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Should not be there. See more comment below.
